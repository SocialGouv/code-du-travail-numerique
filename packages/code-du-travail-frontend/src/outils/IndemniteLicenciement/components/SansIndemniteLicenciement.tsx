import { Container, Title } from "@socialgouv/cdtn-ui";
import React from "react";

import { branches } from "../branches";

type Props = {
  children?: React.ReactNode;
  idcc: string;
};

export default function SansIndemniteLicenciement({
  idcc,
  children,
}: Props): JSX.Element {
  const selectedBranche = branches.find((br) => br.value === idcc)!;

  return (
    <Container>
      <Title as="h3">{selectedBranche.label}</Title>
      <p>
        Pour cette branche, le calcul de l’indemnité de licenciement se base sur
        l’indemnité légale de licenciement.
      </p>
      {children}
    </Container>
  );
}
