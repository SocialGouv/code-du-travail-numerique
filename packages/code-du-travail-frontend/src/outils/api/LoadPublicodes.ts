import { modeles } from "@socialgouv/modeles-social";

export const loadPublicodes = (simulator: string): any => {
  switch (simulator) {
    case "preavis-retraite":
      return modeles;
    case "indemnite-licenciement":
      return modeles;
    default:
      return null;
  }
};
