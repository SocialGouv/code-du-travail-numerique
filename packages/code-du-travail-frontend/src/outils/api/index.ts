import {
  PublicodesSimulator,
  SingletonPublicodesHelper,
} from "@socialgouv/modeles-social";
import { searchAgreement } from "../../conventions/Search/api/agreement.service";
import * as Sentry from "@sentry/nextjs";
import { Agreement } from "@socialgouv/cdtn-utils";

export const loadPublicodes = <T extends PublicodesSimulator>(
  simulator: PublicodesSimulator,
  idcc?: string
) => {
  return SingletonPublicodesHelper.getInstance<T>(simulator, idcc);
};

export const getCc3239Informations = async (): Promise<Agreement> => {
  let defaultInformations = {
    url: "https://www.legifrance.gouv.fr/affichIDCC.do?idConvention=KALICONT000044594539",
    id: "KALICONT000044594539",
    num: 3239,
    shortTitle: "Particuliers employeurs et emploi à domicile",
    slug: "3239-particuliers-employeurs-et-emploi-a-domicile",
    title: "Particuliers employeurs et emploi à domicile",
  } as Agreement;
  try {
    const res = await searchAgreement("3239");
    defaultInformations = res[0];
  } catch (e) {
    console.error(e)
    Sentry.captureException(e);
  }
  return defaultInformations;
};
