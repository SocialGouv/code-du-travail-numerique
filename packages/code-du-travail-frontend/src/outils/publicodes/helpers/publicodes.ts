import Engine from "publicodes";

import { convertDaysIntoBetterUnit, SituationElement } from "..";
import { PublicodesPreavisRetraiteResult } from "../types/preavis-retraite";

export function handleExecutePreavisRetraite(
  engine: Engine,
  situation: SituationElement[],
  rule: string
): PublicodesPreavisRetraiteResult {
  engine.setSituation(buildSituation(situation));
  const result = engine.evaluate(rule);
  return convertDaysIntoBetterUnit(result.nodeValue as unknown as string);
}

export function newSituation(
  engine: Engine,
  situation: SituationElement[],
  targetRule: string,
  args: Record<string, string>
): any {
  // Situation is an array to keep the order of the answers
  const currentSituation = situation;
  const newSituation: SituationElement[] = [];

  // Update the current situation with new values
  currentSituation.forEach((element) => {
    // Keep the data only if always here in the form
    if (args[element.name]) {
      newSituation.push({
        name: element.name,
        rawNode: element.rawNode,
        value: args[element.name],
      });
    }
  });

  // Add the new entries from the form
  Object.entries(args).forEach(([key, value]) => {
    if (!newSituation.find((element) => element.name === key)) {
      const publiKey = key.replace(/ - /g, " . ");
      const detail = engine.getRule(publiKey);
      newSituation.push({
        name: key,
        rawNode: detail.rawNode,
        value: value,
      });
    }
  });

  engine.setSituation(buildSituation(newSituation));
  const result = engine.evaluate(targetRule);

  return {
    missingArgs: buildMissingArgs(engine, result.missingVariables),
    result,
    situation: newSituation,
  };
}

export const buildSituation = (
  map: Array<SituationElement>
): Record<string, string> => {
  const situation: Record<string, string> = {};
  map.forEach((arg) => {
    situation[arg.rawNode.nom] = arg.value;
  });
  return situation;
};

export const buildMissingArgs = (
  engine: Engine,
  missingArgs: Record<string, number>
): Array<any> => {
  return Object.entries(missingArgs)
    .map(([key, value]) => {
      const detail = engine.getRule(key);
      return {
        indice: value,
        name: key.replace(/ \. /g, " - "),
        rawNode: detail.rawNode,
      };
    })
    .sort((a, b) => b.indice - a.indice);
};
