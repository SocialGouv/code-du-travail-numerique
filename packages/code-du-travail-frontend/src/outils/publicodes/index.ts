export * from "./context";
export * from "./helpers";
export * from "./types";
export * from "./usePublicodes";
export * from "./utils";
