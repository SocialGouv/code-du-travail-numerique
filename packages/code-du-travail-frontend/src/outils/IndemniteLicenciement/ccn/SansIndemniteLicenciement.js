import { Container, Title } from "@socialgouv/cdtn-ui";
import React from "react";

import { branches } from "../branches";
import { IndemniteLegale } from "../components/IndemniteLegale";
import { getIndemniteFromFinalForm } from "../indemnite";

function Step({ form }) {
  const {
    values: { branche },
  } = form.getState();

  const { indemniteLegale, infoCalculLegal } = getIndemniteFromFinalForm(form);
  const selectedBranche = branches.find((br) => br.value === branche);

  return (
    <Container>
      <Title as="h3">{selectedBranche.label}</Title>
      <p>
        Pour cette branche, le calcul de l’indemnité de licenciement se base sur
        l’indemnité légale de licenciement.
      </p>
      <IndemniteLegale formValues={form} />
    </Container>
  );
}

export default Step;
