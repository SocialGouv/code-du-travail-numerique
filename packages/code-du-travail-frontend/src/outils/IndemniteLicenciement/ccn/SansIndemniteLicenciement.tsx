import { Container, Title } from "@socialgouv/cdtn-ui";
import React from "react";

import { branches } from "../branches";
import { IndemniteLegale } from "../components/IndemniteLegale";

function Step({ form }) {
  const {
    values: { branche },
  } = form.getState();

  const selectedBranche = branches.find((br) => br.value === branche);

  return (
    <Container>
      {selectedBranche && (
        <>
          <Title as="h3">{selectedBranche.label}</Title>
          <p>
            Pour cette branche, le calcul de l’indemnité de licenciement se base
            sur l’indemnité légale de licenciement.
          </p>
          <IndemniteLegale form={form} />
        </>
      )}
    </Container>
  );
}

export default Step;
