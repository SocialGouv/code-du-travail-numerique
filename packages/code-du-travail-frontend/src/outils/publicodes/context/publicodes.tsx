import { getNotifications, getReferences } from "@socialgouv/modeles-social";
import Engine from "publicodes";
import React, { createContext, useMemo } from "react";

import { handleExecutePreavisRetraite } from "../helpers";
import {
  PublicodesContext as PublicodesContextType,
  PublicodesProviderRule,
  PublicodesSupportedSimulator,
} from "../types";

export const PublicodesContext = createContext<PublicodesContextType | null>(
  null
);

export const PublicodesProvider = ({
  children,
  rules,
  targetRule,
}: PublicodesProviderRule) => {
  const engine = useMemo(() => {
    return new Engine(rules);
  }, [rules]);

  return (
    <PublicodesContext.Provider
      value={{
        execute:
          s === PublicodesSupportedSimulator
            ? handleExecutePreavisRetraite()
            : null,
        getNotifications: () => getNotifications(engine),
        getReferences: () => getReferences(engine),
        missingArgs: null,
        result: null,
        setSituation: null,
        situation: Array.from(data.situation.values()),
      }}
    >
      {children}
    </PublicodesContext.Provider>
  );
};
