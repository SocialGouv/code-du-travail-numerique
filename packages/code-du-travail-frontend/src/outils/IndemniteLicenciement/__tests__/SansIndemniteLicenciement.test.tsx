import { render } from "@testing-library/react";
import React from "react";

import { loadPublicodes } from "../../api/LoadPublicodes";
import { PublicodesProvider, PublicodesSimulator } from "../../publicodes";
import SansIndemniteLicenciement from "../ccn/SansIndemniteLicenciement";

describe("<SansIndemniteLicenciement />", () => {
  it("should render", () => {
    const form = {
      getState() {
        return {
          values: {
            absencePeriods: [],
            anciennete: 6,
            branche: "1518",
            contrat: "cdi",
            dateEntree: "2013-07-02",
            dateNotification: "2019-07-02",
            dateSortie: "2019-07-31",
            fauteGrave: false,
            hasAbsenceProlonge: false,
            hasSameSalaire: true,
            hasTempsPartiel: false,
            inaptitude: false,
            salaire: "2390",
            salairePeriods: [],
            salaires: [],
          },
        };
      },
    };
    const { container } = render(
      <PublicodesProvider
        rules={loadPublicodes("indemnite-licenciement")}
        targetRule="contrat salarié . indemnité de licenciement"
        simulator={PublicodesSimulator.INDEMNITE_LICENCIEMENT}
      >
        <SansIndemniteLicenciement form={form} />
      </PublicodesProvider>
    );
    expect(container).toMatchSnapshot();
  });
});
