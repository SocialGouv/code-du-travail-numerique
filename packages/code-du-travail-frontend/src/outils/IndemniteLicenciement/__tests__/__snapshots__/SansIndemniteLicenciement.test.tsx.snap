// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SansIndemniteLicenciement /> should render 1`] = `
.c0 {
  max-width: 124rem;
  margin: 0 auto;
  padding: 0 2rem;
}

.c3 {
  position: absolute;
  background-color: #7994d4;
  border-radius: 0.6rem;
  top: 50%;
  left: 0;
  width: 0.4rem;
  height: 100%;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
}

.c1 {
  margin-top: 3.2rem;
  margin-bottom: 2rem;
  margin-left: auto;
  text-align: left;
}

.c2 {
  position: relative;
  margin: 0;
  color: #2f3b6c;
  font-weight: normal;
  font-size: 2.8rem;
  font-family: "Merriweather",serif;
  line-height: 1.25;
  padding-left: 3.2rem;
  text-align: left;
}

.c4 {
  margin-top: 3.2rem;
  margin-bottom: 1rem;
  color: #4d73b8;
  font-weight: 600;
  font-size: 1.8rem;
  font-family: "Open Sans",sans-serif;
}

.c5 {
  color: #f66663;
  font-weight: 700;
  font-size: 1.8rem;
  white-space: pre-line;
}

.c6 {
  font-size: 1.5em;
}

@media (max-width:600px) {
  .c0 {
    padding: 0 1rem;
  }
}

@media print {
  .c0 {
    max-width: 100%;
    padding: 0;
  }
}

@media (max-width:600px) {
  .c1 {
    margin-bottom: 1rem;
    margin-left: auto;
  }
}

@media (max-width:600px) {
  .c2 {
    padding-left: 1.6rem;
  }
}

@media (max-width:600px) {
  .c2 {
    font-size: 2.2rem;
  }
}

@media (max-width:600px) {
  .c5 {
    font-size: 1.6rem;
  }
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <h3
        class="c2"
      >
        <span
          class="c3"
        />
        Convention collective nationale de l’Animation
      </h3>
    </div>
    <p>
      Pour cette branche, le calcul de l’indemnité de licenciement se base sur l’indemnité légale de licenciement.
    </p>
    <h2
      class="c4"
    >
      Indemnité légale
    </h2>
    <p>
      Le code du travail prévoit un montant minimum de :
       
      <strong
        class="c5 c6"
      >
        0
         € brut.
      </strong>
       
    </p>
  </div>
</div>
`;
